[MM-53102] Add feature keywords highlight without notification

M-ZubairAhmed opened the pull request:
#### Summary
- Adds a new section under settings/notifications for adding custom multi word keywords that gets highlighted without notification
- Adds a new classname for highlighting words although the styling is same as mentions highlights
- Added few components to ReduxFromProps pattern
- Adds supported type for hook of PluginComponent type
- Add upsel for highlight without notification
- Moved 'setting_item.tsx' to components folder
- Improved prop names and function structure for setting_item, setting_item_max and setting_item_min
- Moved 'toggle_modal_button.tsx' to components folder
- Removed t and utility messages from few components
- Fixed bug where tooltip was not getting rendered on restrictedButtons
- Improved the mobile view of the settings modal
- Adds E2E for the feature

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53102
https://mattermost.atlassian.net/browse/MM-53808
https://mattermost.atlassian.net/browse/MM-55145
https://mattermost.atlassian.net/browse/MM-53480


#### Screenshots

https://github.com/mattermost/mattermost/assets/17708702/08c2e3f5-f42b-42e3-b89b-649c65c97850


#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Added custom keywords with highlights without notification feature.
```

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

mattermost-build commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

marianunez commented on webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:
```diff
@@ -221,6 +221,16 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
         }
 
+        // Add the only highlight without notification keyword
+        if (user.notify_props.highlight_keys) {
+            keys = keys.concat(user.notify_props.highlight_keys.split(',').map((key) => {
+                return {
+                    key,
+                    caseSensitive: true,
+                };
+            }));
+        }
```
This seems to work to highlight the words but it seems that this `getCurrentUserMentionKeys` is also called to list in the RHS actions when opening the Mentions RHS. I don't think we want to consider these highlights as mentions given they are not being notified about it

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

hmhealey commented on webapp/channels/src/components/markdown/index.ts:
```diff
@@ -11,27 +13,79 @@ import {getPost} from 'mattermost-redux/selectors/entities/posts';
+    highlightKeys?: UserHighlightKey[];
     postId?: string;
+
+    /*
```
I'm liking the `PropsFromRedux` pattern here, but I feel like it'd be better to leave the rest of these props in the component file instead of moving them here. That does mean that we might have to repeat a few fields when a component has an `OwnProps` field, but I think that's fine since most components don't use `OwnProps`. I think `connect` is actually a bit less performant if you use `ownProps` because it has to rerun `mapStateToProps` even if `state` doesn't change then.

Could you move the rest of these fields back to the `Props` type in the component file?

M-ZubairAhmed replied:
done moved to the component

hmhealey commented on webapp/channels/src/components/post_markdown/index.ts:
```diff
@@ -54,10 +54,30 @@ export function makeGetMentionKeysForPost(): (
-    mentionKeys: MentionKey[];
+export type OwnProps = {
+
+    /**
```
Same here

hmhealey commented on webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:
```diff
@@ -223,6 +223,27 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
 
+export type UserHighlightKey = {
+    key: string;
+    caseSensitive?: boolean;
```
Does having this field make using the existing code in `text_formatting.ts` easier, or can we leave it out? I can't imagine people will ever want this to be case sensitive, but there's one case that we need case sensitive mentions (just for first names and just because some people will be named words like Will that overlap with common words)

M-ZubairAhmed replied:
removed case sensitive key

hmhealey commented on webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:
```diff
@@ -223,6 +223,27 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
+
+        if (user.notify_props.highlight_keys.length > 0) {
+            user.notify_props.highlight_keys.split(',').forEach((key) => {
+                highlightKeys.push({key, caseSensitive: true});
```
Speaking of that, I think we'd actually want these highlight keys to not be case sensitive

johndavidlugtu replied:
+1 on keys not being case sensitive

cc: @abhijit-singh

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

marianunez reviewed (changes requested):
Haven't done a full code review yet but it seems that the webapp is crashing when opening the user settings modal with 
```
 Uncaught TypeError: Cannot read properties of undefined (reading 'length')
    at St (user_settings_notifications.tsx:173:52
 ```

abhijit-singh commented:
The screencap looks good but facing the same issue as Maria. Will check again once that is fixed.

abhijit-singh reviewed (approved):
Looks awesome, thanks @M-ZubairAhmed!

yasserfaraazkhan commented:
Hi @M-ZubairAhmed , 
putting down a test in zephyr  https://github.com/mattermost/mattermost-test-management/pull/119/files

while trying out the changes observed a behaviour,

- if user has multiple phrases, that begin with  same words, not all phrases get highlighted 
in below screenshots, we can see only the first phrase was being highlighted.


<img width="466" alt="Screenshot 2023-10-05 at 11 35 08 PM" src="https://github.com/mattermost/mattermost/assets/15979783/0615144f-3fa9-4a54-b272-92a6be7d8e8d">

<img width="301" alt="Screenshot 2023-10-05 at 11 35 14 PM" src="https://github.com/mattermost/mattermost/assets/15979783/eae04985-ea02-44a2-906d-260cc4691052">

marianunez commented on webapp/channels/src/i18n/en.json:
```diff
@@ -5430,6 +5430,10 @@
+  "user.settings.notifications.keywordsWithHighlight.extraInfo": "These keywords will be shown to you with a highlight when anyone sends a message that includes them.",
+  "user.settings.notifications.keywordsWithHighlight.inputTitle": "Enter non case-sensitive keywords, press Tab or use commas to separate them:",
+  "user.settings.notifications.keywordsWithHighlight.none": "None",
+  "user.settings.notifications.keywordsWithHighlight.title": "Keywords That Get Highlighted (without notifications)",
```
@abhijit-singh seems there is inconsistency in casing with this title. Should the words within the parenthesis also be in capitals?

marianunez reviewed (changes requested):
@M-ZubairAhmed overall code LGTM - however, requesting changes due to the ongoing conversation regarding moving this to a Professional feature

yasserfaraazkhan commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

yasserfaraazkhan commented:
/e2e-tests

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/381653) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/381653)

yasserfaraazkhan reviewed (changes requested):
- when admin opens up `Start trial` or `purchase modal`, the notification modal stays on top.
- special character are not highlighted. (this is expected, right ?)
- IF a sentence ends with `DOT` or character, its not highlighted. ( is this expected ?)

<img width="500" alt="image" src="https://github.com/mattermost/mattermost/assets/15979783/f17506a3-5d25-41d0-a0d9-c97931586212">

https://github.com/mattermost/mattermost/assets/15979783/3b6f325d-fcab-4660-a8a2-f9af4835a006

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
@abhijit-singh 
> In the default state, there should not be a grey background on the section. The grey background should only be there on the hover state, just like for other items.

Changed


> Can we trigger the Professional feature tooltip when the user hovers anywhere on the entire section instead of just on the professional tag?

In the code the restricted button encapsulates all the logic for license check etc and contains the tooltip as well, so it would be hard to bring out the tooltip from that to the item. If this is not too much of a blocker can we skip this? It matches the current behavior of user groups in the global header menu item.


> In the default view, can we truncate the text that shows what has been set if it overflows, similar to how it happens for Keywords That Trigger Notification

Done

> Can we make the capitalization consistent for both the sections around keywords

Done

> I also noticed that email IDs also don't get highlighted in this new functionality

Added

> when admin opens up Start trial or purchase modal, the notification modal stays on top

Fixed

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

marianunez commented on webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:
```diff
@@ -147,9 +150,10 @@ function getDefaultStateFromProps(props: Props): State {
 
     if (props.user.notify_props) {
-        if (props.user.notify_props.mention_keys) {
+        if (props.user.notify_props.mention_keys.length > 0) {
```
Should we add a null check here in case just in case? Before it was checking directly if it existed ? Also, why this change?

```suggestion
        if (props.user.notify_props.mention_keys?.length > 0) {
```

M-ZubairAhmed replied:
changed

marianunez commented on webapp/channels/src/utils/text_formatting.test.ts:
```diff
@@ -304,6 +306,101 @@ describe('highlightCurrentMentions', () => {
+    test('should replace highlight keywords with tokens', () => {
+        const text = 'This is a test message with some keywords';
+        const tokens = new Map();
+        const highlightKeys = [
+            {key: 'keywords'},
+            {key: 'test'},
+        ];
```
Should we include phrases here with spaces as well for this test?

M-ZubairAhmed replied:
added a test

marianunez reviewed (approved):
Great work @M-ZubairAhmed! 🚀 Minor nits below

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
@abhijit-singh @yasserfaraazkhan gentle ping on this one

M-ZubairAhmed commented:
/update-branch

yasserfaraazkhan reviewed (commented):
self-hosted:
- tested with uploading license. With sysadmin and user's account. On desktop and web
- tested with instance being on Enterprise trial
@johndavidlugtu @M-ZubairAhmed @marianunez 
If the Trial has ended, should the customer be able to use the highlighted keywords? 
It can also be a scenario that which the admin uploads a trial license and removes it. In between users have added keywords. These get highlighted.

<img width="600" alt="image" src="https://github.com/mattermost/mattermost/assets/15979783/706990e0-5e19-4f57-bbff-07816ec2e0b6">

abhijit-singh reviewed (approved):
Looks good to me, thanks for the great work and all the iterations on this @M-ZubairAhmed! 

To @yasserfaraazkhan's point - 
> If the Trial has ended, should the customer be able to use the highlighted keywords?

I believe we should stop highlighting keywords in this case but if possible maintain the keywords data added by the user. If that's complex to do, I'd vote for getting rid of the keywords altogether and disable the feature if they don't have the required license.

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
@yasserfaraazkhan 

> If the license is E10 and E20, feature is enabled. But when saving the words. we see this error.

As per docs, both the licences are on the end of life https://docs.mattermost.com/about/editions-and-offerings.html#other-mattermost-plans

Added the fix for the highlight words not to get highlighted after the trial period.

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
> Hi @M-ZubairAhmed the fixes work well now.
> 
> I was testing the admin notification part. It shows users requested and on which plan is available, but the feature name is missing. Please let me know this would be a follow up fix.
> 
> 

Thank you for pointing this one out. I thought the markdown came from the server but i was wrong this custom render is in webapp

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/386178) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/386178)

M-ZubairAhmed commented:
/update-branch

yasserfaraazkhan reviewed (approved):
Thank you for the changes @M-ZubairAhmed 
the notify admin part also works as expected.

<img width="500" alt="image" src="https://github.com/mattermost/mattermost/assets/15979783/ffcff4e3-97c0-4857-af23-7893b23160a8">

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/386531) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/386531)

M-ZubairAhmed commented:
@johndavidlugtu can you please review

M-ZubairAhmed commented:
/update-branch

hmhealey commented on webapp/channels/src/components/user_settings/notifications/index.ts:
```diff
@@ -18,11 +22,27 @@ const mapStateToProps = (state: GlobalState) => {
         enableAutoResponder,
         isCollapsedThreadsEnabled: isCollapsedThreadsEnabled(state),
         isCallsRingingEnabled: isCallsEnabled(state, '0.17.0') && isCallsRingingEnabledOnServer(state),
+        isStarterFree,
```
I guess my previous comments apply here too

M-ZubairAhmed replied:
yes added selector

hmhealey commented on webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:
```diff
@@ -145,9 +148,10 @@ function getDefaultStateFromProps(props: Props): State {
 
     if (props.user.notify_props) {
-        if (props.user.notify_props.mention_keys) {
+        if (props.user.notify_props?.mention_keys?.length > 0) {
```
I don't think these `?.` are needed since we null checked `props.user.notify_props` above

M-ZubairAhmed replied:
Yeah once the app crashed on new user because the prop `highlight_keys` was never set for it, so i added these checks

hmhealey replied:
I think we needed to check for `highlight_keys`  specifically because it didn't previously exist in the `UserNotifyProps`, but I don't think we need to null check `notify_props` or `mention_keys`

hmhealey commented on webapp/channels/src/components/widgets/menu/menu_items/restricted_indicator.tsx:
```diff
@@ -107,27 +110,29 @@ const RestrictedIndicator = ({
-                        {icon}
-                        {ctaExtraContent}
-                    </ToggleModalButton>
+                    <span>
```
Do we need this extra span?

M-ZubairAhmed replied:
yes it is https://github.com/mattermost/mattermost/pull/24905#discussion_r1360788142

hmhealey commented on webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:
```diff
@@ -224,6 +224,26 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
+    (user: UserProfile) => {
+        const highlightKeys: HighlightWithoutNotificationKey[] = [];
+
+        if (user?.notify_props?.highlight_keys?.length > 0) {
```
```suggestion
        if (user.notify_props.highlight_keys?.length > 0) {
```

M-ZubairAhmed replied:
changed

hmhealey commented on webapp/channels/src/utils/text_formatting.tsx:
```diff
@@ -704,6 +715,79 @@ export function highlightCurrentMentions(
+    }
+
+    highlightKeys.
+        sort((a, b) => b.key.length - a.key.length).
```
I forget where we landed on fixing handling overlapping highlights since we discussed that last week. Did we decide this method of fixing it is enough for now?

M-ZubairAhmed replied:
yes will open a ticket for the case

johndavidlugtu commented:
<img width="584" alt="Screenshot 2023-11-10 at 10 38 07 AM" src="https://github.com/mattermost/mattermost/assets/101303489/106f6a5d-5568-49cb-84f2-0d5214a82f2a">

[Admin view] Can we update the second sentence of the body to, "Upgrade to Professional plan to unlock this feature."


<img width="584" alt="Screenshot 2023-11-10 at 10 43 14 AM" src="https://github.com/mattermost/mattermost/assets/101303489/2f10bf3f-09c7-455a-84d0-12efd950ac64">
[End user view] Change second sentence of the body to, "Request your admin to upgrade to Mattermost Professional to unlock this feature."

@M-ZubairAhmed really great work here!

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/386924) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/386924)

M-ZubairAhmed commented on webapp/channels/src/components/post_markdown/post_markdown.test.tsx:
```diff
@@ -22,6 +24,14 @@ describe('components/PostMarkdown', () => {
+        isUserCanManageMembers: false,
+        isStarterFree: true,
+        isEnterpriseReady: false,
+        dispatch: jest.fn(),
```
Yes, I rechecked it's due to `makeMapStateToProps` along with PropsWithRedux.

M-ZubairAhmed commented on webapp/channels/src/components/post_markdown/index.ts:
```diff
@@ -54,30 +56,46 @@ export function makeGetMentionKeysForPost(): (
             isMilitaryTime: getBool(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.USE_MILITARY_TIME, false),
             timezone: getCurrentTimezone(state),
             hideGuestTags: getConfig(state).HideGuestTags === 'true',
+            isStarterFree,
+            isEnterpriseReady,
```
We need a distinction between an empty array and one with no license for scenarios.

hmhealey replied:
I'm a bit surprised that this code would need it, but I can see that the setting screen would. It's totally fine as-is though

M-ZubairAhmed commented:
> <img alt="Screenshot 2023-11-10 at 10 38 07 AM" width="584" src="https://user-images.githubusercontent.com/101303489/282105439-106f6a5d-5568-49cb-84f2-0d5214a82f2a.png">
> [Admin view] Can we update the second sentence of the body to, "Upgrade to Professional plan to unlock this feature."
> 
> @M-ZubairAhmed really great work here!
> 
> Pre-approving assuming the copy will be fixed, thanks

Changed

M-ZubairAhmed commented:
/e2e-cancel

mattermost-build commented:
Looks like nothing had to be canceled.

M-ZubairAhmed commented:
/update-branch

hmhealey commented on webapp/channels/src/utils/license_utils.ts:
```diff
@@ -103,3 +104,14 @@ export const licenseSKUWithFirstLetterCapitalized = (license: ClientLicense) =>
     return sku.charAt(0).toUpperCase() + sku.slice(1);
 };
+
+export function isEnterpriseOrCloudOrSKUStarterFree(license: ClientLicense, subscriptionProduct: Product | undefined, isEnterpriseReady: boolean) {
```
0/5 but since the license, product, and `isEnterpriseReady` all come from redux, I think we could turn `isEnterpriseOrCloudOrSKUStarterFree` into a full selector that gets all of those from redux.

Also, I'm still trying to understand what we're trying to do by checking for these. Are we trying to determine if we can upsell the user on Enterprise?

M-ZubairAhmed replied:
this checks if we are on a starter licence, if yes then we are checking if this is an enterprise ready app then we are adding an upsel

hmhealey reviewed (approved):
One more comment on the code that checks the license, but it's nothing blocking, so this looks good to me

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/387111) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/387111)

M-ZubairAhmed commented:
> Successfully triggered E2E testing! [GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/387111) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/387111)

Only 1 failure not related to this change
